include/watchdog.h

```diff
@@ -1,116 +1,168 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
- *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
-
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
-};
+/**
+ * @brief Watchdog operational mode.
+ */
+#define WDT_OP_MODE_RESET		(0)
+#define WDT_OP_MODE_INTERRUPT_RESET	BIT(0)
+#define WDT_OP_MODE_MASK		(0x1)
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog behavior when CPU is in sleep state.
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
-};
+#define WDT_NO_PAUSE_IN_SLEEP	(0)
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+#define WDT_PAUSE_IN_SLEEP_MASK (1 << 1)
 
+/**
+ * @brief Watchdog behavior when CPU is halted by the debugger.
+ */
+#define WDT_NO_PAUSE_HALTED_BY_DBG	(0)
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+#define WDT_PAUSE_HALTED_BY_DBG_MASK	(1 << 2)
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog configuration struct.
+ *
+ * timeouts - pointer to an array storing timeouts values in microseconds
+ *	      for all enabled channels. If watchdog supports multiple channels
+ *	      but only one counter reload value then the first value from the
+ *	      array will be taken.
+ * interrupt_fn - function that is going to be called if
+ *		  WDT_OP_MODE_INTERRUPT_RESET is enabled.
+ * operation - a bitfield with following parts:
+ *    operational mode    [ 0 ]       - reset or interrupt -> reset
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
+ * num_of_channels - number of channels that should be enabled on
+ *		     given watchdog instance.
  */
 struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+	u32_t *timeouts;
```

Comment:
timeouts with an 's'?

Reply:
@nvlsianpu No, there is no way to get the real timeout. Driver implementation should approximate to the closest possible value or maybe to the closest higher value? `get_config` will be probably removed as no one can find any use case for it right now. 
Maybe *timeouts table should be in/out parameter, after leaving the function it would store the real set values?

Reply:
@m-kru Works for me as well.